src/core/mod.rs

```diff
@@ -18,654 +18,665 @@
+				mean.mul_f64(delay.min(10.0))
+			})
+	}
+
+	/// Either generate and return a cover packet or pop and return the packet at the head of one
+	/// of the authored packet queues. May return `None` if cover packets are disabled, we fail to
+	/// generate a cover packet, or there are no active sessions (though in the no active sessions
+	/// case `next_authored_packet_delay` should return `None` and so this function should not
+	/// really be called).
+	pub fn pop_next_authored_packet(&mut self, ns: &dyn NetworkStatus) -> Option<AddressedPacket> {
+		// This function should be called according to a Poisson process. Randomly choosing between
+		// sessions and cover kinds here is equivalent to there being multiple independent Poisson
+		// processes; see https://www.randomservices.org/random/poisson/Splitting.html
+		let mut rng = rand::thread_rng();
 
-	/// Should be called when a peer is disconnected.
-	pub fn remove_connected_peer(&mut self, id: &MixPeerId) {
-		self.connected_peers.remove(id);
-	}
+		// First pick the session
+		let sessions: ArrayVec<_, 2> = self
+			.sessions
+			.enumerate_mut()
+			.filter(|(rel_session_index, _)| {
+				self.session_status.phase.gen_cover_packets(*rel_session_index)
+			})
+			.collect();
+		let (rel_session_index, session) = match sessions.into_inner() {
+			Ok(sessions) => {
+				// Both sessions active. We choose randomly based on their rates.
+				let periods = sessions
+					// TODO This could be replaced with .each_ref() once it is stabilised, allowing
+					// the collect/into_inner/expect at the end to be dropped
+					.iter()
+					.map(|(_, session)| session.mean_authored_packet_period.as_secs_f64())
+					.collect::<ArrayVec<_, 2>>()
+					.into_inner()
+					.expect("Input is array of length 2");
+				let [session_0, session_1] = sessions;
+				if rng.gen_bool(periods[0] / (periods[0] + periods[1])) {
+					session_0
+				} else {
+					session_1
+				}
+			},
+			// Either just one active session or no active sessions. This function shouldn't really
+			// be called in the latter case, as next_authored_packet_delay() should return None.
+			Err(mut sessions) => sessions.pop()?,
+		};
 
-	fn cover_message(&mut self) -> Option<(MixPeerId, Packet)> {
-		let mut rng = rand::thread_rng();
-		let message = fragment::Fragment::create_cover_fragment(&mut rng);
-		let (id, key) = self.random_cover_path()?;
+		self.invalidated |= Invalidated::NEXT_AUTHORED_PACKET_DEADLINE;
 
-		let hop = sphinx::PathHop { id, public_key: key.into() };
-		let (packet, _no_surb) =
-			sphinx::new_packet(&mut rng, vec![hop], message.into_vec(), None).ok()?;
-		Some((id, packet))
+		// Choose randomly between drop and loop cover packet
+		if rng.gen_bool(self.config.loop_cover_proportion) {
+			gen_cover_packet(&mut rng, &session.topology, ns, CoverKind::Loop, &self.config)
+		} else {
+			self.session_status
+				.phase
+				.allow_requests_and_replies(rel_session_index)
+				.then(|| session.authored_packet_queue.pop())
```

Comment:
Should the authored packet be more prioritary than the Loop covers?

Reply:
Hmm not sure. It might make sense to do that. I had requests and replies replacing drop cover packets because they are more like drop packets -- they go to another node on the network. However requests can result in replies, which perhaps makes them more similar to loop cover packets, although the timing is really not the same (loop cover packets will come back much sooner than replies on average). Loop cover packets may also be used for monitoring the network, in which case we wouldn't really want to replace them. They're not currently used for this of course. In Loopix, I believe loop cover packets are never replaced.